tests/functional/conservation/conserv.py

```diff
@@ -0,0 +1,289 @@
+    )
+
+    ElectronModel(closure="isothermal", Te=0.)
+
+    sim = ph.global_vars.sim
+
+    timestamps = np.arange(0, sim.final_time, 50*sim.time_step)
+
+    for quantity in ["B"]:
+        ElectromagDiagnostics(
+            quantity=quantity,
+            write_timestamps=timestamps,
+            compute_timestamps=timestamps,
+        )
+
+    for name in ["domain", "levelGhost", "patchGhost"]:
+        ParticleDiagnostics(quantity=name,
+                            compute_timestamps=timestamps,
+                            write_timestamps=timestamps,
+                            population_name="protons")
+
+
+
+#############################################################
+#               POSTPROCESS ROUTINES
+#############################################################
+#plotting et al. functions
+def get_times(path):
+    import h5py
+    f = h5py.File(path, 'r')
+    times = np.array(sorted([float(s.strip("t")) for s in list(f.keys())]))
+    f.close()
+    return times
+
+
+
+def kinetic_energy(particles, kind="iso"):
+    """
+    return the total kinetic energy of given particles
+    assuming parallel is x, and perp is y,z
+    This also assumes protons, so mass is 1.
+    """
+    if kind=="iso":
+        return 0.5*np.sum((particles.v[:,0]**2 +\
+                           particles.v[:,1]**2 +\
+                           particles.v[:,2]**2)* particles.weights)
+    if kind=="perp":
+        return 0.5*np.sum((particles.v[:,1]**2 +\
+                           particles.v[:,2]**2)* particles.weights)
+    if kind=="para":
+        return 0.5*np.sum((particles.v[:,0]**2)* particles.weights)
+
+
+
+def total_particles(parts, fun, lvlNbr=0, **kwargs):
+    """
+    given a particle hierarchy, calculate a total
+    quantity on a given level, quantity being estimated
+    by the callback "fun" (could be kinetic energy, momentum, et.)
+    """
+    for ilvl, lvl in parts.patch_levels.items():
+        if lvlNbr == ilvl:
+            tot = 0.
+            for ip, patch in enumerate(lvl.patches):
+                keys = list(patch.patch_datas.keys())
+                pdata = patch.patch_datas[keys[0]]
+                particles= pdata.dataset
+                per_patch = fun(particles, **kwargs)
+                tot += per_patch
+    return tot
+
+
+
+def total_kinetic(parts, lvlNbr=0, kind="iso"):
+    return total_particles(parts, kinetic_energy, lvlNbr, kind=kind)
+
+
+
+def mag_energy(B, lvlNbr=0):
+    """
+    return the total magnetic energy on a given level
+    """
+    for ilvl, lvl in B.patch_levels.items():
+        if lvlNbr == ilvl:
+            tot = 0.
+            for ip, patch in enumerate(lvl.patches):
+                pdata = patch.patch_datas["Bx"]
+
+                # Bx is primal, but By and Bz dual so
+                # we average By and Bz onto Bx to use them
+                # in the same formula
+                bx = patch.patch_datas["Bx"].dataset[5:-5]
```

Comment:
there are no reason why we could change the 5 ghosts cells, but for clearness, may be these `5` and `4` should be a constant variable named `GHOST_NBR`

Reply:
should be accessible from https://github.com/PHAREHUB/PHARE/blob/master/pyphare/pyphare/pharesee/hierarchy.py#L100

    patch.patch_datas["Bx"].ghost_nbr[0]